### PR TITLE
Use distgen for memcached

### PR DESCRIPTION
**Pull request by phracek**
This Pull Request implements distgen.

@bkabrda @TomasTomecek Please look at it and review it. Thanks.

### CODE REVIEW

File: specs/configuration.yml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+maintainer: Petr Hracek <phracek@redhat.com>
+
+expose: 11211
+
+description:  Memcached is High Performance, Distributed Memory Object Cache.
+
+user: 1000
+
+envvars:
```
Comment (TomasTomecek) [inline specs/configuration.yml]: no need to group those, right?

File: Dockerfile [hunk 2]
```diff
@@ -1,37 +1,38 @@
...
 # Ports:
 #  * 11211
 
-ENV NAME=memcached ARCH=x86_64
-LABEL MAINTAINER "Petr Hracek" <phracek@redhat.com>
+ENV NAME={{ spec.envvars.name }} \
+    VERSION={{ spec.envvars.version }} \
+    RELEASE={{ spec.envvars.release }} \
+    ARCH={{ spec.envvars.arch }}
+LABEL {{ spec.maintainer }}
```
Comment (TomasTomecek) [inline Dockerfile]: this won't work, check how this expands

File: specs/multispec.yml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+specs:
+  distroinfo:
+    fedora:
+      distros:
+        - fedora-26-x86_64
+        - fedora-25-x86_64
+      vendor: "Fedora Project"
+      source_url: "some.url.fedoraproject.org"
+      distro_specific_help: "Some Fedora specific help"
```
Comment (TomasTomecek) [inline specs/multispec.yml]: hm? what's the point of doing this? this is not an example repo, this is a real image

File: Makefile [hunk 4]
```diff
@@ -1,14 +1,37 @@
...
-build:
-	docker build --tag=$(IMAGE_NAME) .
+DG_EXEC = ${DG} --distro ${DISTRO}.yaml --spec specs/configuration.yml --multispec specs/multispec.yml --multispec-selector version=${VERSION}
+DISTRO_ID = $(shell ${DG_EXEC} --template "{{ config.os.id }}")
 
 run: build
 	docker run -d $(IMAGE_NAME)
 
+debug: build
+	docker run -t -i $(IMAGE_OPTIONS) $(IMAGE_NAME) bash
```
Comment (TomasTomecek) [inline Makefile]: don't you wanna enable the debugging env var?

File: Makefile [hunk 5]
```diff
@@ -1,14 +1,37 @@
-.PHONY: build run defult
+.PHONY: build run debug test doc
 
 IMAGE_NAME = modularitycontainers/memcached
+IMAGE_OPTIONS = \
+    -p 127.0.0.1:11211:11211
 
-default: run
+VERSION = 1.14
```
Comment (TomasTomecek) [inline Makefile]: since this does not build from a module, you can't be sure that you'll be getting 1.14, is that true?

**Review by TomasTomecek — CHANGES_REQUESTED** [review]
@bkabrda your turn

I think we should talk about the default distgen template -- what we want to support and how it should look

File: Makefile (hunk 5 above)
Comment (phracek) [reply Makefile]: @bkabrda I wrote the version against memcached -v value. Is it correct? Like in bind case https://github.com/container-images/bind/blob/master/specs/multispec.yml#L19 .

Comment (TomasTomecek) [reply Makefile]: I would personally prefer [the way I did it](https://github.com/container-images/tools/pull/4/files#diff-37c04ce6822c2b1d4aa6db755e21f3d2R9) -- don't mention versioning but instead use whatever the platform offers, so:

```
variants:
  upstream:
  fedora:
  centos:
  ...
```

But I would prefer discussing this.

File: Dockerfile [hunk 6]
```diff
@@ -1,35 +1,36 @@
-FROM registry.fedoraproject.org/f26-modular/boltron
+FROM {{ config.docker.from }}
```
Comment (bkabrda) [inline Dockerfile]: You probably want to use `{{ config.docker.registry }}/{{ config.docker.from }}`

Comment (bkabrda) [reply Dockerfile]: Also, `{{ config.docker.from }}` will give you `fedora:26`, which is something that you didn't have before. You may want to consider if that's what you want to do.

File: Makefile (hunk 5 above)
Comment (bkabrda) [reply Makefile]: I guess this comes down to what the author wants/needs. In this case, I don't see `{{ spec.version }}` used anywhere, so I'm wondering why this (and the `version` section in multispec) is even necessary.

**Review by bkabrda — CHANGES_REQUESTED** [review]

File: Makefile (hunk 5 above)
Comment (phracek) [reply Makefile]: @bkabrda @TomasTomecek New distgen version fix the problem with `--multispec-selector`.

File: tests/Makefile [hunk 7]
```diff
@@ -1,6 +1,3 @@
-MODULE_LINT=/usr/share/moduleframework/tools/modulelint.py
```
Comment (TomasTomecek) [inline tests/Makefile:1]: I don't understand what's the point of having such makefile.

File: Dockerfile [hunk 8]
```diff
@@ -1,35 +1,36 @@
...
 # Ports:
 #  * 11211
 
-ENV NAME=memcached ARCH=x86_64
-LABEL MAINTAINER "Petr Hracek" <phracek@redhat.com>
+ENV NAME={{ spec.envvars.name }} \
+    VERSION={{ spec.envvars.version }} \
+    RELEASE={{ spec.envvars.release }} \
+    ARCH={{ spec.envvars.arch }}
+MAINTAINER {{ spec.maintainer }}
```
Comment (TomasTomecek) [inline Dockerfile]: `LABEL maintainer`, `MAINTAINER` instruction is deprecated

File: Dockerfile [hunk 9]
```diff
@@ -1,35 +1,36 @@
...
 LABEL summary="High Performance, Distributed Memory Object Cache" \
     name="$FGC/$NAME" \
     version="0" \
     release="1.$DISTTAG" \
     architecture="$ARCH" \
     com.redhat.component=$NAME \
-    usage="docker run -p 11211:11211 docker.io/modularitycontainers/memcached" \
+    usage="docker run -p 127.0.0.1:{{ spec.expose }}:{{ spec.expose }} docker.io/modularitycontainers/memcached" \
```
Comment (TomasTomecek) [inline Dockerfile]: image name here should be templated

File: Makefile [hunk 10]
```diff
@@ -1,14 +1,36 @@
-.PHONY: build run defult
+.PHONY: build run debug test doc
 
 IMAGE_NAME = modularitycontainers/memcached
+IMAGE_OPTIONS = \
+    -p 127.0.0.1:11211:11211
```
Comment (TomasTomecek) [inline Makefile]: why localhost?

File: help/help.md [hunk 11]
```diff
@@ -39,6 +39,8 @@ Lists of security-related attributes that are opened to the host.
 -p 11211:11211
     Opens container port 11211 and maps it to the same port on the host.
 
+{{ spec.distro_specific_help }}
```
Comment (TomasTomecek) [inline help/help.md]: there is no such value in spec

File: specs/multispec.yml [hunk 12]
```diff
@@ -0,0 +1,12 @@
+version: 1
+
+specs:
+  distroinfo:
+    fedora:
```
Comment (TomasTomecek) [inline specs/multispec.yml:5]: we also need upstream

**Review by TomasTomecek — CHANGES_REQUESTED** [review]

File: tests/Makefile (hunk 7 above)
Comment (phracek) [reply tests/Makefile:1]: MTF provides a binary mtf which can call tests directly. https://github.com/fedora-modularity/meta-test-family/pull/101 Now we can delete Makefile.

Comment (TomasTomecek) [reply tests/Makefile:1]: \o/

File: Dockerfile [hunk 13]
```diff
@@ -1,35 +1,36 @@
...
     name="$FGC/$NAME" \
     version="0" \
     release="1.$DISTTAG" \
     architecture="$ARCH" \
     com.redhat.component=$NAME \
-    usage="docker run -p 11211:11211 docker.io/modularitycontainers/memcached" \
-    help="Runs memcached, which listens on port 11211. No dependencies. See Help File below for more details." \
-    description="memcached is a high-performance, distributed memory object caching system, generic in nature, but intended for use in speeding up dynamic web applications by alleviating database load." \
-    io.k8s.description="memcached is a high-performance, distributed memory object caching system, generic in nature, but intended for use in speeding up dynamic web applications by alleviating database load." \
+    usage="docker run -p 127.0.0.1:{{ spec.expose }}:{{ spec.expose }} docker.io/modularitycontainers/{{ spec.envvars.name }}" \
```
Comment (TomasTomecek) [inline Dockerfile]: `-p 127.0.0.1:...`, why?

File: Makefile [hunk 14]
```diff
@@ -1,14 +1,45 @@
-.PHONY: build run defult
+.PHONY: build run default debug test doc upstream downstream
 
 IMAGE_NAME = modularitycontainers/memcached
```
Comment (TomasTomecek) [inline Makefile]: this should be dynamic: but depends whether you want to use this makefile in downstream

File: specs/multispec.yml (hunk 12 above)
Comment (TomasTomecek) [reply specs/multispec.yml]: resolved with variants

File: specs/multispec.yml [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+  variant:
+    upstream:
+      image_name: modularitycontainers/memcached
+      readme_description: |-
+        [![Build Status](https://travis-ci.org/container-images/memcached.svg?branch=master)](https://travis-ci.org/container-images/memcached)
+
+        There is a detailed information about this image available in its [help
+        page](https://github.com/container-images/memcached/blob/master/help/help.md).
+    downstream:
+      image_name: f26/memcached
```
Comment (TomasTomecek) [inline specs/multispec.yml]: this is not correct:
1. registry is missing
2. f26 should probably come from the spec

**Review by TomasTomecek — CHANGES_REQUESTED** [review]
getting closer

File: specs/multispec.yml (hunk 12 above)
Comment (phracek) [reply specs/multispec.yml:5]: Upstream is incorporated.

**Comment by TomasTomecek** [comment]
Can you please also include `.travis.yml` and [hooks/pre_build](https://github.com/bkabrda/container-image-template-dg/blob/master/hooks/pre_build) so we're sure it builds fine?

File: .travis.yml [hunk 16]
```diff
@@ -0,0 +1,12 @@
...
+  - docker
+script:
+  - hooks/pre_build
+  # Docker Hub hack
+  - sudo cp -av ./Dockerfile.template ./Dockerfile
+  - make build
+env:
+  - DG_BINARY="docker run -v $(pwd):/var/dgdir slavek/distgen"
+notifications:
+email: false
```
Comment (TomasTomecek) [inline .travis.yml]: `\t`

File: hooks/pre_build [hunk 17]
```diff
@@ -0,0 +1,12 @@
+image='docker.io/slavek/distgen'
```
Comment (TomasTomecek) [inline hooks/pre_build]: This file needs to be executable.

File: Dockerfile [hunk 18]
```diff
@@ -1,35 +1,36 @@
...
 # Ports:
 #  * 11211
 
-ENV NAME=memcached ARCH=x86_64
-LABEL MAINTAINER "Petr Hracek" <phracek@redhat.com>
+ENV NAME={{ spec.envvars.name }} \
+    VERSION={{ spec.envvars.version }} \
+    RELEASE={{ spec.envvars.release }} \
+    ARCH={{ spec.envvars.arch }}
+LABEL MAINTAINER {{ spec.maintainer }}
```
Comment (TomasTomecek) [inline Dockerfile]: all our labels should be lower case

File: help/help.md [hunk 19]
```diff
@@ -17,7 +17,7 @@ Files added to the container during docker build include: /files/memcached.sh
 # USAGE
 To get the memcached container image on your local system, run the following:
 
-    docker pull hub.docker.io/phracek/memcached 
+    docker pull hub.docker.io/phracek/{{ spec.envvars.name }}
```
Comment (TomasTomecek) [inline help/help.md]: docker.io/modularitycontainers/...

File: help/help.md [hunk 20]
```diff
@@ -3,7 +3,7 @@
 % February 6, 2017
 
 # NAME
-memcached - High Performance, Distributed Memory Object Cache.
+{{ spec.envvars.name }} - {{ spec.description }}
 
 # DESCRIPTION
 Memcached is a high-performance, distributed memory object caching system, generic in nature, but intended for use in speeding up dynamic web applications by alleviating database load.
```
Comment (TomasTomecek) [inline help/help.md:9]: two lines below there is a mention of fedora/24, that's not accurate

**Review by TomasTomecek — CHANGES_REQUESTED** [review]
build passed, just a couple of small issues and we are good to merge!

**Comment by TomasTomecek** [comment]
CI is green, let's merge

**Pull request merged**